In [1]:
# The code was removed by Watson Studio for sharing.

# Segmenting and Clustering Neighbourhoods in Toronto

## Part 1: Scraping and preprocessing postal code data

#### Obtaining postal data

We will be obtaining Toronto postal code data from the following table https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M and converting it to a Pandas Dataframe

In [2]:
import pandas as pd
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'


In [3]:
#Use Pandas to read the first table on the webpage into a dataframe 
df_postalcodes = pd.read_html(url)[0]


In [4]:
#Inspect top rows of dataframe
df_postalcodes.head(10)

Postal Code           Borough                                Neighbourhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                 Not assigned
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge

Lets see the initial shape of the table before we refine it

In [5]:
df_postalcodes.shape

(180, 3)

It initially has 180 records and 3 columns

#### Preprocessing postal data

The requirements state to only process rows that are assigned to boroughs, so we will drop all "Not assigned" rows

In [6]:
df_postalcodes = df_postalcodes[df_postalcodes['Borough'] != "Not assigned"]

In [7]:
df_postalcodes

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

The requirements state to group every postal code that is assigned to multiple neighbourhoods together, seperated by comas. So we will find out how many rows contain the same postal code.

In [8]:
occurances = df_postalcodes['Postal Code'].value_counts()

for index, val in occurances.items():
    if(val > 1):
        print(index, " appears more than once")

Running this code we find there are no occurances of postal codes appearing on more than once. Inspecting the table further reveals the wikipedia page has been updated to group all neighbourhoods with the same postal code together.

Next, we are to assign the borough name to the Neighbourhood column if the Neighbourhood is not assigned. Lets create a boolean mask of all rows with unassigned Neighbourhoods and filter the dataframe to see if there are any occurances.

In [9]:
occurances = df_postalcodes['Neighbourhood'] == 'Not assigned'
df_postalcodes[occurances]

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

The output table is empty, indicating there are no postal codes without assigned neighbourhoods left in the dataframe and no further processing is required. 

Finally, lets see the shape of the refined table

In [10]:
df_postalcodes.shape

(103, 3)

There are 103 records and 3 columns in the table.

## Part 2: Obtaining location data

We will now use the Geocoder Python package to get lattitude and longitude coordinates of each neighbourhood

In [11]:
!pip install geocoder # install geocoder library
import geocoder # import geocoder

We will loop through each postal code in the dataframe and make API calls to the geocoder service. Due to the unreliable nature of the geocoder API it sometimes takes multiple calls to recieve a response and therefore we will loop for each code until a response is recieved.


In [12]:
lat = []
long = []

for code in df_postalcodes['Postal Code']:
    lat_lng_coords = None
    
    while (lat_lng_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(code))
        lat_lng_coords = g.latlng
        #print(code, g.latlng)

    lat.append(lat_lng_coords[0])
    long.append(lat_lng_coords[1])


Now lets append those lattitude and longitude lists to the dataframe

In [13]:
df_postalcodes['Latitude'] = lat
df_postalcodes['Longitude'] = long

df_postalcodes.head()

Postal Code           Borough                                Neighbourhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

   Latitude  Longitude  
2  43.75245  -79.32991  
3  43.73057  -79.31306  
4  43.65512  -79.36264  
5  43.72327  -79.45042  
6  43.66253  -79.39188

#### Exporting location data for future use

I'm going to need this data for the next capstone project, so lets export it. Extra steps needed to be taken in order to save the DataFrame to the IBM Cloud, in addition to the typical use of Pandas to_cvs() method. 

I needed to create a project Access Token using the steps outlined here: https://dataplatform.cloud.ibm.com/docs/content/wsj/analyze-data/project-lib-python.html#save-data 

The cell below contains my IBM Watson project credentials and therefore must be hidden, but this is the code used:

```python
from project_lib import Project
project = Project(project_id='myprojectid', project_access_token='myaccesstoken')
```

*Note: it would not allow me to pass in credentials as string variables*

In [14]:
# The code was removed by Watson Studio for sharing.

In [15]:
project.save_data(data=df_postalcodes.to_csv(index=False),file_name='toronto_location_data.csv',overwrite=True)

{'file_name': 'toronto_location_data.csv',
 'message': 'File saved to project storage.',
 'bucket_name': 'datavsciencecapstoneproject-donotdelete-pr-ajhmjtmpgksd8h',
 'asset_id': '8f6fd023-d0a3-40a7-9afb-f027f7f2d051'}

The data is now ready to be used in the next project. 

## Part 3: Clustering neighbourhoods


In this section we will utilize the FourSquare API to request data of the different venues in each neighbourhood, and then use KNN clustering to group each neighbourhood into categories. 

In [16]:
import numpy as np # library to handle data in a vectorized manner

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment if not installed
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!pip install folium
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment if not installed
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


#### Mapping

Using the geopy library, we will obtain the coordinates of Toronto

In [17]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


We will create a map centered on Toronto and superimpose markers for each neighbourhood

In [18]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, Neighbourhood in zip(df_postalcodes['Latitude'], df_postalcodes['Longitude'], df_postalcodes['Borough'], df_postalcodes['Neighbourhood']):
    label = '{}, {}'.format(Neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Using FourSquare API

We will use the FourSquare API to obtain all the venues within a 500m radius around the center of each group of neighbourhoods (by postal code)

My API credentials are hidden in the cell below, for security

In [19]:
# The code was removed by Watson Studio for sharing.

In [20]:
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

We will make a test API call to see the JSON object returned

In [21]:
radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, 43.75245, -79.32991, VERSION, radius, LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fd040bab63c403432c65735'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.756950004500005,
    'lng': -79.3236918238658},
   'sw': {'lat': 43.7479499955, 'lng': -79.3361281761342}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': [{

Lets define a function to extract venue type from results

In [22]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Clean JSON data

In [23]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = pd.json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

lets now create a function to repeat this for each neighbourhood

In [24]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

And now we shall run it

In [25]:
toronto_venues = getNearbyVenues(names = df_postalcodes['Neighbourhood'], latitudes = df_postalcodes['Latitude'],longitudes = df_postalcodes['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Now lets observe the resulting dataframe

In [26]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods                43.75245                -79.32991   
1         Parkwoods                43.75245                -79.32991   
2  Victoria Village                43.73057                -79.31306   
3  Victoria Village                43.73057                -79.31306   
4  Victoria Village                43.73057                -79.31306   

                Venue  Venue Latitude  Venue Longitude     Venue Category  
0     Brookbanks Park       43.751976       -79.332140               Park  
1       Variety Store       43.751974       -79.333114  Food & Drink Shop  
2        Wigmore Park       43.731023       -79.310771               Park  
3  Memories of Africa       43.726602       -79.312427      Grocery Store  
4       Guardian Drug       43.730584       -79.307432           Pharmacy

And lets see how many of each venue are in each group of neighbourhoods

In [27]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Agincourt                                                               16   
Alderwood, Long Branch                                                   4   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East                                       19   
Berczy Park                                                             60   
Birch Cliff, Cliffside West                                              4   
Brockton, Parkdale Village, Exhibition Place                            84   
Business reply mail Processing Centre, South Ce...                     100   
CN Tower, King and Spadina, Railway Lands, Harb...                      76   
Caledonia-Fairbanks                                                      7   
Canada Post Gateway Processing Centre                                  100   
Cedarbrae                                                                1   
Central Bay Street                                                      76   
Christie                                                                11   
Church and Wellesley                                                    79   
Clarks Corners, Tam O'Shanter, Sullivan                                 16   
Cliffside, Cliffcrest, Scarborough Village West                          8   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              26   
Davisville North                                                         8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                         4   
Don Mills                                                               15   
Dorset Park, Wexford Heights, Scarborough Town ...                       2   
Downsview                                                               41   
Dufferin, Dovercourt Village                                            17   
East Toronto, Broadview North (Old East York)                            3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                       6   
Fairview, Henry Farm, Oriole                                            55   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          2   
Garden District, Ryerson                                               100   
Glencairn                                                               13   
Golden Mile, Clairlea, Oakridge                                         10   
Guildwood, Morningside, West Hill                                        3   
Harbourfront East, Union Station, Toronto Islands                       58   
High Park, The Junction South                                            4   
Hillcrest Village                                                        2   
Humber Summit                                                            2   
Humberlea, Emery                                                         4   
Humewood-Cedarvale                                                       5   
India Bazaar, The Beaches West                                          18   
Islington Avenue, Humber Valley Village                                  8   
Kennedy Park, Ionview, East Birchmount Park                              3   
Kensington Market, Chinatown, Grange Park                               57   
Kingsview Village, St. Phillips, Martin Grove G...                       5   
Lawrence Manor, Lawrence Heights                                        51   
Lawrence Park                                                            2   
Leaside                                                                 33   
Little Portugal, Trinity                                                45   
Malvern, Rouge                                                           3

### Analyising Data

Now lets briefly look at some of the trends in the dataset

We will use one-hot encoding to prepare for averaging the frequency of venue type in each Neighbourhood

In [28]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# Move Neighbourhood column to first index
first_col = toronto_venues.get('Neighbourhood')
#toronto_onehot = toronto_onehot.drop(columns=['Neighbourhood'])
toronto_onehot.insert(0, 'Neighbourhood', first_col)

toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  \
0         Parkwoods                  0                  0   
1         Parkwoods                  0                  0   
2  Victoria Village                  0                  0   
3  Victoria Village                  0                  0   
4  Victoria Village                  0                  0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   

   Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  \
0            0              0                0          0           0   
1            0              0                0          0           0   
2            0              0                0          0           0   
3            0              0                0          0           0   
4            0              0                0          0           0   

   Badminton Court  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0                0           0       0     0    0               0   
1                0           0       0     0    0               0   
2                0           0       0     0    0               0   
3                0           0       0     0    0               0   
4                0           0       0     0    0               0   

   Basketball Stadium  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                   0         0           0                   0          0   
1                   0         0           0                   0          0   
2                   0         0           0                   0          0   
3                   0         0           0                   0          0   
4                   0         0           0                   0          0   

   Bike Trail  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0       0              0          0         0   
1           0       0              0          0         0   
2           0       0              0          0         0   
3           0       0              0          0         0   
4           0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridge  Bubble Tea Shop  \
0                     0               0        0       0                0   
1                     0               0        0       0                0   
2                     0               0        0       0                0   
3                     0               0        0       0                0   
4                     0               0        0       0                0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
0             0              0         0            0         0   
1             0              0         0            0         0   
2             0              0         0            0         0   
3             0              0         0            0         0   
4             0              0         0            0         0   

   Business Service  Butcher  Café  Camera Store  Candy Store  \
0                 0        0     0             0            0   
1              

Lets take the venue frequency mean for each group of neighbourhoods

In [ ]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Now lets view the 5 most frequent venues in each group to get a feel for how they could potentially be classified

In [ ]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

#### Observations

Lets look at the first 3 areas that appear alphabetically and observe some trends

1. Agincourt has many Asian resturants so it might be an area with many Asian residents
2. Alderwood/Long Branch looks like an urban area with convenidence stores, pubs, gyms, and performing arts venues
3. Bayview Village has many golfing ranges, trails, parks, and construction areas indicating it is a more rural area that is under developement 


Lets sort the venues in each neighbourhood in decending order

Here is a function to accomplish that task

In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Lets create a new dataframe to store the Neighbourhoods and their venues sorted by frequency

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
Neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
Neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    Neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

Neighbourhoods_venues_sorted.head()

### Clustering Each Neighbourhood

To begin, lets determine the optimal k value

In [ ]:
import matplotlib.pyplot as plt


toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)


inertia=[]

for k in range(2, 15):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(toronto_grouped_clustering)
    inertia.append(kmeans.inertia_)

plt.plot(range(2,15), inertia, 'bx-')
plt.xlabel('k')
plt.ylabel('Sum_of_squared_distances')
plt.title('Elbow Method For Optimal k')
plt.show()


The elbow is at k = 6, so lets use that

In [ ]:
# set number of clusters
kclusters = 6


# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100] 

In [ ]:
toronto_grouped_clustering

Merge postal code data with venue data

In [ ]:
# add clustering labels
Neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_postalcodes

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each Neighbourhood
toronto_merged = toronto_merged.join(Neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged # check the last columns!

#### Removing NaN rows

If you look through the dataframe you will notice some rows have NaN values. Turns out the FourSquare API didn't have any data for venues within 500m of those areas. So we will simply remove those rows.

In [ ]:
toronto_merged = toronto_merged.dropna()

#### Visualizing Clusters

In [ ]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Classifications

Now that we have fitted the neighbourhoods into 6 different clusters using KNN, we must now determine what each classification actually represents by observing the patterns of venues in each class.

##### Cluster 0

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

It appears like cluster 0 are rural areas with lots of farms, parks, fields, etc.

##### Cluster 1

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Cluster 1 clearly makes up the bulk of Toronto neighbourhoods, and it is made up of very urban areas. The most frequent venues are coffee shops,  shopping districts, resturaunts, etc.

##### Cluster 2

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Cluster 2 appears to be suburban with parks, resturants, and schools. Farms are also fairly frequent in these areas. 

##### Cluster 3

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Cluster 3 is apparently an outliner with no similar places like it. Strangely it looks very similar to cluster 2. Perhaps the high frequency of trails and zoo exhibits threw the KNN off?

##### Cluster 4

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Cluster 4 is yet another apparent outlier, with the high auto garage frequency probably throwing the KNN calculation off. Once again it looks like cluster two. 

##### Cluster 5

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Finally, cluster 5 contains 3 similar neighbourhoods which appear to be suburban.

### Conclusion

Analysis has revealed the neighbourhoods in Toronto are best classified into 3 major groups: Urban, Suburban, and Rural. There are no further apparent distinctions that can split them into sub-categories. 